pkg/credentials/filter/ring_buffer.go

```diff
@@ -0,0 +1,85 @@
+package filter
+
+import (
+	"math"
+)
+
+// RingBuffer implements a simple ring buffer with a static maximal length.
+type RingBuffer struct {
+	start         int
+	end           int
+	currentLength int
+	buffer        []byte
+}
+
+// NewRingBuffer creates a RingBuffer with the given length.
+func NewRingBuffer(length int) *RingBuffer {
```

Comment:
You need to keep the last N bytes somewhere to check if they contain a secret value and only print out stuff that is clearly no secret. In our case that means everything which was not detected as secret in the ring buffer and drops out when something is added to the ring buffer. As all secret values are known, the maximum ring buffer size will be the maximum secret length.

Reply:
I dont know if there is a good implementation out there that fits the bill here. But there seems to be nothing in the standard lib.

Reply:
To check my understanding, the entrypoint is replacing the default stdout/stderr buffers with this new filtered credentials writer? would it be possible to just have the whole string in memory and check if it contains secret values before writing to stdout/stderr? obviously this holds a lot more data in memory.. I'm just trying to brainstorm a bit because I think adding a new buffer implementation adds quite a bit of complexity and we should do it only if it's really necessary.

Reply:
Having the whole string in memory would mean that your output is delayed until the run is finished, which is bad for long running jobs where you want the logs timely to get a status of the running job.
This implementation on the other hand will print out everything except the newest N bytes directly.

Reply:
Great point. Would the Reader and Writer from bufio work? there's a Peek function (https://pkg.go.dev/bufio#Reader.Peek) that could be used to check if the next n bytes contain a secret

Reply:
The bufio package just implements buffered readers and writers on top of other readers and writers. That itself is not a ringbuffer implementation. A ringbuffer ist just an array that you treat as a circular buffer by jumping from end to start regularly. Its super efficient because you do not need to allocate new arrays when your buffer runs full. I doubt that a general buffered reader implementation does that. 

As I said I do not know any implementation in the standard lib. There are some OS packages like the following:
https://github.com/sahmad98/go-ringbuffer
https://github.com/smallnest/ringbuffer

But the first is too simplistic and the second even uses mutexes which we do not need here I think. Also you would make the whole implementation more "dependent" on a very small package which is perhaps not even cared for anymore which is kind of bad for these few lines of code.

Reply:
I agree that we don't want to add a new dependency like the one you listed, but this still adds complexity to an already large PR. Would it be possible to either use a plain slice of bytes as a buffer, or move the ring buffer implementation into its own PR?

Reply:
I have splitted out the ring buffer into a separate pull request with more tests to cover it.

https://github.com/tektoncd/pipeline/pull/5990